Source/Particles/PhysicalParticleContainer.cpp

```diff
@@ -622,17 +622,138 @@ PhysicalParticleContainer::AddPlasmaFromFile(ParticleReal q_tot,
         }
     } // IO Processor
     auto const np = particle_z.size();
-    AddNParticles(0, np,
-                  particle_x.dataPtr(),  particle_y.dataPtr(),  particle_z.dataPtr(),
-                  particle_ux.dataPtr(), particle_uy.dataPtr(), particle_uz.dataPtr(),
-                  1, particle_w.dataPtr(),1);
+    InitializeRuntimeAttributesAndAddNParticles (0, np,
+                particle_x.dataPtr(),  particle_y.dataPtr(),  particle_z.dataPtr(),
+                particle_ux.dataPtr(), particle_uy.dataPtr(), particle_uz.dataPtr(),
+                particle_w.dataPtr(), 1, -1, amrex::RandomEngine{});
 #endif // WARPX_USE_OPENPMD
 
     ignore_unused(q_tot, z_shift);
 
     return;
 }
 
+void
+PhysicalParticleContainer::InitializeRuntimeAttributesAndAddNParticles (
```

Comment:
Could this code setting up the attributes be incorporated into `AddNParticles`? This would avoid having extra copies of the attribute arrays (which could be problematic with large numbers of particles). Also, after these changes, is `AddNParticles` called directly anywhere else?

Reply:
`AddNParticles` can directly be called from Python - we make heavy use of it for our thermionic injectors. So my vote would be to not absorb too much extra functionality into `AddNParticles`.

Reply:
Ok, presumably you are setting the needed attributes from Python. But would this have the same issue being address here, that some attributes may have incorrect default values? Also, this means that you couldn't use the `addRealAttributes` and `addIntegerAttributes` features. Perhaps the routine `InitializeRuntimeAttributesAndAddNParticles` could be made callable from Python also?

Reply:
What I was thinking of was modifying AddNParticles to use the default values as being setup here in `InitializeRuntimeAttributesAndAddNParticles` for any attributes that are not passed in. This would not affect your current use from Python.

Reply:
Yes, I did not want to directly modify the `AddNParticles` function because I saw that it was used from Python. It is also still used when splitting particles in `PhysicalParticleContainer::SplitParticles`, possibly incorrectly with respect to runtime parameters (I don't know this function well enough but in this case we may need to copy the runtime parameters from the initial particles so the `InitializeRuntimeAttributesAndAddNParticles` may not be good).

Also, this PR indeed does not fix the issue when using Python but I wanted to fix the pure C++ version first because QED (which is the module for which the issue is the most dangerous IMO) does not yet work with Python, so I thought a Python fix was less urgent.

What you suggest in your latest message @dpgrote is that we modify `AddNParticles` so that it accepts a list of runtime parameters that are passed in `attr_real` and "default initializes" (same as I did here) the ones that are not passed in? If so, I think that it's generally a good idea. We'd have to slightly modify the Python interface so that `AddNParticles` can know which attributes are passed and we'd also have to make sure that we don't break the `SplitParticles` routine (but it may already be broken). I think this can be done either here or in a follow-up PR, what do you think?

Reply:
Yes, that is what I am suggesting. I make the suggestion since I think it will simplify the code some and make it more robust.

If I understand it correctly, currently `nattr` attributes are passed into `AddNParticles` and any beyond that are set to zero. Perhaps this PR could fix so that a proper default is used instead of only zero. A second PR could update the Python interface to allow more flexibility.